src/app/Login/Login.tsx

```diff
@@ -46,49 +46,48 @@ import { Base64 } from 'js-base64';
 import { BasicAuthDescriptionText, BasicAuthForm } from './BasicAuthForm';
 import { BearerAuthDescriptionText, BearerAuthForm } from './BearerAuthForm';
 
-export const Login = (props) => {
+export const Login = () => {
   const serviceContext = React.useContext(ServiceContext);
   const notifications = React.useContext(NotificationsContext);
-
-  const [token, setToken] = React.useState('');
   const [authMethod, setAuthMethod] = React.useState('');
   const addSubscription = useSubscriptions();
-  const onLoginSuccess = props.onLoginSuccess;
 
   const checkAuth = React.useCallback((token, authMethod, userSubmission = false) => {
     let tok = token;
+
     if (authMethod === 'Basic') {
-      tok = Base64.encodeURL(token);
+      tok = Base64.encodeURL(tok);
     } // else this is Bearer auth and the token is sent as-is
+
     addSubscription(
-      serviceContext.api.checkAuth(tok, authMethod)
+      serviceContext.login.checkAuth(tok, authMethod)
       .pipe(first())
       .subscribe(v => {
         if (v) {
-          onLoginSuccess();
+          //what goes here?
         } else if (userSubmission) {
           notifications.danger('Authentication Failure', `${authMethod} authentication failed`);
         }
       })
     );
-  }, [serviceContext, serviceContext.api, addSubscription, onLoginSuccess, notifications, authMethod]);
+  }, [serviceContext, serviceContext.login, addSubscription, notifications, authMethod]);
 
   const handleSubmit = React.useCallback((evt, token, authMethod) => {
-    setToken(token);
     setAuthMethod(authMethod);
     checkAuth(token, authMethod, true);
     evt.preventDefault();
-  }, [setToken, setAuthMethod, checkAuth]);
+  }, [serviceContext, serviceContext.login, setAuthMethod, checkAuth]);
+
 
   React.useEffect(() => {
-    const sub = serviceContext.api.getAuthMethod().subscribe(setAuthMethod);
+    const sub = serviceContext.login.getAuthMethod().subscribe(setAuthMethod);
     checkAuth('', 'Basic'); // check auth once at component load to query the server's auth method
     return () => sub.unsubscribe();
-  }, [serviceContext, serviceContext.api, setAuthMethod, checkAuth]);
+  }, [serviceContext, serviceContext.login, setAuthMethod, checkAuth]);
 
   React.useEffect(() => {
     const sub =
-      combineLatest(serviceContext.api.getToken(), serviceContext.api.getAuthMethod(), timer(0, 5000))
+      combineLatest(serviceContext.login.getToken(), serviceContext.login.getAuthMethod(), timer(0, 5000))
```

Comment:
When I go to localhost:9000, I can see that four requests, `/auth`, '`v1/recordings', `v1/targets`, and '/auth`, fail with a `401`. The console log tells me that the `Alert` component was trying to render a JSON object in the `{message}` with the text `Error: Unauthorized`, which prevents the page from rendering.

If I comment out the `{message}`, the page will endlessly call `/auth` in sets of two. I think at least one of the `useEffect` hooks is triggering repeatedly.

Why are there two `/auth` calls at the start? Why are both `serviceContext` and `serviceContext.login` dependencies to the `useEffect` hooks?

![image](https://user-images.githubusercontent.com/84587295/131713778-01a4a487-98e9-4766-b340-58a71bccefcb.png)

Reply:
There is one `/auth` call performed unconditionally at application start-up, which it uses to probe the server for the expected auth method (Bearer or Basic). The server responds with a 401 and a header indicating the expected auth method. Once that's known the application can attempt to actually auth the user with the credentials they supply, using the now-known method, so that actual attempt would be the second `/auth` that you see.

The `useEffect` dependency arrays contain references to everything that would "invalidate" the closure of the lambda function passed in if the reference changed, so that React can re-render and recompute the lambda with the updated references when they change, basically. `ServiceContext` and the `LoginService` are both actually "static" in the sense that there is only ever 1 instance for the lifetime of the application, so these should never cause a re-render to occur in practice. Not including them in the array leads to linter warnings, however, and there is no harm in including them.